comms/dht/src/network_discovery/discovering.rs

```diff
@@ -188,44 +214,170 @@ impl Discovering {
+        Ok(peer_stream)
+    }
+
+    async fn get_peer_response(
+        &mut self,
+        stream: &mut ClientStreaming<GetPeersResponse>,
+        sync_peer: &NodeId,
+    ) -> Result<Option<Result<GetPeersResponse, RpcStatus>>, NetworkDiscoveryError> {
+        let rpc_streaming_timeout = self.config().network_discovery.bootstrap_rpc_streaming_timeout;
+        match tokio::time::timeout(rpc_streaming_timeout, stream.next()).await {
+            Ok(response) => Ok(response),
+            Err(_) => Err(NetworkDiscoveryError::Timeout {
+                operation: "get_peer_response".to_string(),
+                peer: sync_peer.to_hex(),
+                duration: format!("{:.2?}", rpc_streaming_timeout),
+            }),
+        }
+    }
+
+    async fn request_peers(&mut self, sync_peer: &NodeId, client: rpc::DhtClient) -> Result<(), NetworkDiscoveryError> {
+        debug!(
+            target: LOG_TARGET,
+            "Discovering: Requesting {} peers from `{}`",
+            self.params.num_peers_to_request,
+            sync_peer
+        );
+        let mut stream = self.get_stream(client, sync_peer).await?;
         let mut counter = 0;
+        let mut invalid_peer_signatures = 0;
+        let peer_sync_immediate_ban_threshold = self.config().network_discovery.peer_sync_immediate_ban_threshold;
         #[allow(clippy::mutable_key_type)]
         let mut peers_received = HashSet::new();
-        while let Some(resp) = stream.next().await {
+        while let Some(resp) = self.get_peer_response(&mut stream, sync_peer).await? {
             counter += 1;
             if counter > self.params.num_peers_to_request {
-                warn!(target: LOG_TARGET, "Remote peer sent more peers than we requested.");
+                warn!(target: LOG_TARGET, "Discovering: Sync peer `{}` sent more peers than we requested.", sync_peer);
                 return Err(NetworkDiscoveryError::TooManyPeersReceived);
             }
-            let GetPeersResponse { peer } = resp?;
+            if invalid_peer_signatures >= peer_sync_immediate_ban_threshold {
+                warn!(
+                      target: LOG_TARGET,
+                     "Discovering: Too many ban-able offences ({} of {}) received from peer `{}`. Stopping further \
+                     processing.",
+                      peer_sync_immediate_ban_threshold, counter, sync_peer
+                );
+                return Err(NetworkDiscoveryError::BanableOffencesWhileSyncing(
+                    invalid_peer_signatures,
+                ));
+            }
+            let GetPeersResponse { peer } = match resp {
+                Ok(val) => val,
+                Err(err) => {
+                    warn!(
+                        target: LOG_TARGET,
+                        "Discovering: Sync peer `{}` sent invalid response: {:?}",
+                        sync_peer, err
+                    );
+                    return Err(NetworkDiscoveryError::from(err));
+                },
```

Comment:
```suggestion
            let GetPeersResponse { peer } = resp.inspect_err(|e|
                    warn!(
                        target: LOG_TARGET,
                        "Discovering: Sync peer `{}` sent invalid response: {:?}",
                        sync_peer, err
                    ))?
```

Reply:
Maybe I am missing something here, but the compiler does not allow it because the `inspect_err` method is designed for inspecting errors without consuming the Result, and in each case I need to return on error?

Reply:
hence the `?` at the end?